Fix #7 issue. Improve text validation checking if the text to translate is a number or a URL.

pedrovgs opened the pull request:
Hi 47Deg developers!

In this PR I've changed your text validation policy in ``ClipboardServicesComponentImpl`` to fix the issue #7.

Take a look to the new conditions I've added. Now, if the text to translate is a number or a URL the text is not consider as valid. I've added some tests to ``ClipboardServiceComponentSpec``.

I've used a regular expression to check if the text to translate is a URL or not and the ``forall`` functional combinator to check if the text is a number. 

To write this tests I've used ``clipboardServices.isValidCall shouldEqual false`` but I'm not sure if this is the correct way to assert this expected value :S This is my first PR to a Scala project :)

Maybe all the code related to the text validation should be extracted to another class. But I think for this PR is not needed.

juanpedromoreno commented on src/test/scala/com/fortysevendeg/translatebubble/modules/clipboard/ClipboardServiceComponentSpec.scala:
```diff
@@ -59,6 +59,55 @@ class ClipboardServiceComponentSpec
+      val text = "This is not a number"
+      mockClipItem.getText returns text
+
+      clipboardServices.isValidCall shouldEqual true
```
is `clipboardServices.isValidCall must beTrue` you were looking for? ;)

pedrovgs replied:
yes :) I've updated this PR to use ``must beTrue``. Thanks for the clue.

juanpedromoreno commented on src/main/scala/com/fortysevendeg/translatebubble/modules/clipboard/impl/ClipboardServicesComponentImpl.scala:
```diff
@@ -103,6 +103,18 @@ trait ClipboardServicesComponentImpl
+    !isTrimmedTextEmpty(text) && !isTextANumber(text) && !isTextAUrl(text)
+  }
+
+  private def isTrimmedTextEmpty(text: CharSequence): Boolean = text.toString.trim.length == 0
```
text.toString.trim.isEmpty would be another way

pedrovgs replied:
This is the previous implementation. I've just moved the code to a private function. If you suggest me any better approach to implement this function I can change it :)

juanpedromoreno commented:
Minor comments from my side. Maybe @javipacheco wants take a look as well ;)

Excellent and thanks for your contribution @pedrovgs!

juanpedromoreno commented on src/main/scala/com/fortysevendeg/translatebubble/modules/clipboard/impl/ClipboardServicesComponentImpl.scala:
```diff
@@ -61,7 +61,7 @@ trait ClipboardServicesComponentImpl
 
       result match {
-        case Some(text) if text.toString.trim.length > 0 => lastDate = currentMillis; currentInterval > millisInterval
+        case Some(text) if isValidText(text) => lastDate = currentMillis; currentInterval > millisInterval
```
Your code is perfect but as I know is you first Scala contribution and I going to be more `tiki-tiki` but it's just for your scala knowledge ;) . Of course, it could be out of the scope of this PR but I hope would be useful.
Instead of dealing with the CharSequence in the private functions, as alternative we could pass as a parameter the string directly.

`case Some(text) if isValidText(text.toString)`

juanpedromoreno commented on src/main/scala/com/fortysevendeg/translatebubble/modules/clipboard/impl/ClipboardServicesComponentImpl.scala:
```diff
@@ -103,6 +103,18 @@ trait ClipboardServicesComponentImpl
   }
 
+  private def isValidText(text: CharSequence): Boolean = {
+    !isTrimmedTextEmpty(text) && !isTextANumber(text) && !isTextAUrl(text)
```
Looks fine for you if we use something like this?

```
  private def isValidText(text: String): Boolean =
    text.trim.nonEmpty && !isTextANumber(text) && !isTextAUrl(text)
```
Note: braces are not needed here ;)

juanpedromoreno commented:
As I said before, thanks @pedrovgs !

pedrovgs commented:
You are welcome @juanpedromoreno !! I've added a new commit to the PR with your suggestions. Now the private methods use String as parameter instead of CharSequence and I've removed one method as you suggested :)

franciscodr commented on src/test/scala/com/fortysevendeg/translatebubble/modules/clipboard/ClipboardServiceComponentSpec.scala:
```diff
@@ -59,6 +59,55 @@ class ClipboardServiceComponentSpec
+      clipboardServices.isValidCall must beFalse
+    }
+
+    "ClipboardService should validate text if the text value is not a URL" in new ClipboardMocks {
```
This unit test is very similar to the first one. Maybe you could unify both unit tests and use the following description: `ClipboardService should validate text if the text value is not a number neither an url`.

pedrovgs replied:
If I unify these unit tests I had to add two asserts, two arrange and two acts to the merged test you suggest. I prefer to use two tests in this case :)

franciscodr replied:
I'm sorry but I think that my explanation wasn't very clear. I was referring to remove one of them and change the description of the another one to `ClipboardService should validate text if the text value is not a number neither an url` since the code of both unit test is identical. But nevermind, we could leave the two unit tests too ;)

pedrovgs replied:
Ah!!!!!!!!! Now I understand you xD I'm going to change it and update this PR ;)

pedrovgs replied:
Done, thanks!

franciscodr replied:
Thanks to you and sorry for the confusing comment!

franciscodr commented:
Thank you for taking the time to improve the project @pedrovgs. Minor comments from my side, good job and I hope you have enjoyed with Scala ;)

javipacheco commented:
LGTM! Welcome to Scala!!

![gif](https://lh4.googleusercontent.com/-gaw25jwPfKw/U5lpUUZn2SI/AAAAAAAAuIY/kU6-kfCsztc/w1000-h478-no/14%2B-%2B1.gif)

pedrovgs commented:
@javipacheco your :+1: is the second one :) Can we merge this PR?

javipacheco commented:
@franciscodr will do it today. I only add GIFs :-p

pedrovgs commented:
Thanks!!!!

If you have any other issue in this repository or another one related to Scala where I can help, please let me know :)

![ScalaRules](http://www.reactiongifs.com/r/a-b-yay.gif)
